worlds/kh1/Rules.py

```diff
@@ -0,0 +1,1962 @@
+
+def has_torn_pages(state: CollectionState, player: int, pages_required: int) -> bool:
+    return state.count_from_list_unique(TORN_PAGES, player) >= pages_required
+
+def has_all_arts(state: CollectionState, player: int) -> bool:
+    return state.has_all({"Fire Arts", "Blizzard Arts", "Thunder Arts", "Cure Arts", "Gravity Arts", "Stop Arts", "Aero Arts"}, player)
+
+def has_all_summons(state: CollectionState, player: int) -> bool:
+    return state.has_all({"Simba", "Bambi", "Genie", "Dumbo", "Mushu", "Tinker Bell"}, player)
+
+def has_all_magic_lvx(state: CollectionState, player: int, level) -> bool:
+    return state.has_all_counts({
+        "Progressive Fire": level,
+        "Progressive Blizzard": level,
+        "Progressive Thunder": level,
+        "Progressive Cure": level,
+        "Progressive Gravity": level,
+        "Progressive Aero": level,
+        "Progressive Stop": level}, player)
+
+def has_offensive_magic(state: CollectionState, player: int) -> bool:
+    return state.has_any({"Progressive Fire", "Progressive Blizzard", "Progressive Thunder", "Progressive Gravity", "Progressive Stop"}, player)
+
+def has_reports(state: CollectionState, player: int, eotw_required_reports: int) -> bool:
+    return state.has_group_unique("Reports", player, eotw_required_reports)
+
+def has_final_rest_door(state: CollectionState, player: int, final_rest_door_requirement: str, final_rest_door_required_reports: int, keyblades_unlock_chests: bool):
+    if final_rest_door_requirement == "reports":
+        return state.has_group_unique("Reports", player, final_rest_door_required_reports)
+    if final_rest_door_requirement == "puppies":
+        return has_puppies(state, player, 99)
+    if final_rest_door_requirement == "postcards":
+        return state.has("Postcard", player, 10)
+    if final_rest_door_requirement == "superbosses":
+        return (
+                state.has_all({"Olympus Coliseum", "Neverland", "Agrabah", "Hollow Bastion", "Green Trinity", "Phil Cup", "Pegasus Cup", "Hercules Cup", "Entry Pass"}, player)
+                and has_emblems(state, player, keyblades_unlock_chests)
+                and has_all_magic_lvx(state, player, 2)
+                and has_defensive_tools(state, player)
+                and has_x_worlds(state, player, 7, keyblades_unlock_chests)
+            )
+
+def has_defensive_tools(state: CollectionState, player: int) -> bool:
+    return (
+            (
+                state.has("Progressive Cure", player, 2)
+                and state.has("Leaf Bracer", player)
+                and state.has("Dodge Roll", player)
+            )
+            and (
+                state.has("Second Chance", player)
+                or state.has("MP Rage", player)
+                or state.has("Progressive Aero", player, 2)
+            )
+        )
+
+def has_keyblade(state: CollectionState, player: int, keyblade_required: bool, item: str) -> bool:
+    return not keyblade_required or state.has(item, player)
```

Comment:
I think I would've preferred here if you had done something more like this:

```
world.get_location("Traverse Town 1st District Candle Puzzle Chest").access_rule =\
    lambda state: state.has("Progressive Blizzard", player) # Leave out keyblade stuff here
...

# Add the keyblade rules to all relevant locations if the option is on
if options.keyblades_unlock_chests:
    add_rule(world.get_location("Traverse Town 1st District Candle Puzzle Chest"), 
             lambda state: state.has("Lionheart", player))
    add_rule(world.get_location("Traverse Town 1st District Accessory Shop Roof Chest"),
             lambda state: state.has("Lionheart", player))
    ...
```

This seems like the perfect use case for `add_rule`, and I believe it would be faster-or-equal in all cases to do this, plus more readable in my opinion